meilisearch/meilisearch-swift
# Error handler

## Error handler draft

This is a draft to make the errors more clear.
In order to achieve this, I use the error body from MeiliSearch to feed the error object.

## Test this branch
Because `apiKey` does not work #39, we will use that bug to test our error handler.


1. Run a MeiliSearch with a apikey
```
docker run -it --rm -p 7700:7700 getmeili/meilisearch:v0.13.0rc0 ./meilisearch  --no-analytics=true --master-key=masterKey
```

2. Run the tests 
```bash
swift test
```
3. Observe the output from the DocumentsTests

```bash
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:126: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : failed - The operation couldn’t be completed. (MeiliSearch.MSError error 1.)
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:129: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : Asynchronous wait failed: Exceeded timeout of 1 seconds, with unfulfilled expectations: "Add or replace Movies document".
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:142: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : failed - The operation couldn’t be completed. (MeiliSearch.MSError error 1.)
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:145: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : Asynchronous wait failed: Exceeded timeout of 3 seconds, with unfulfilled expectations: "Add or replace Movies document".
```

In the case of `addDocuments` from the `testAddAndGetDocuments`, the error comes back to `addOrReplaceDocuments`
but does not comes back to the testing method.

When printing the error in `addOrReplaceDocuments` this is the output: 
```bash
MSError(data: Optional(MeiliSearch.MSErrorResponse(message: "You must have an authorization token", errorCode: "missing_authorization_header", errorType: "authentication_error", errorLink: Optional("https://docs.meilisearch.com/errors#missing_authorization_header"))), underlying: Error Domain=HttpStatus Code=401 "(null)")
```

## Expected results
I would like the MSError to appear in the failed message from the tests.

```bash
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:126: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : failed - MSError(data: Optional(MeiliSearch.MSErrorResponse(message: "You must have an authorization token", errorCode: "missing_authorization_header", errorType: "authentication_error", errorLink: Optional("https://docs.meilisearch.com/errors#missing_authorization_header"))), underlying: Error Domain=HttpStatus Code=401 "(null)")
/meili/meilisearch-swift/Tests/MeiliSearchTests/DocumentsTests.swift:129: error: -[MeiliSearchTests.DocumentsTests testAddAndGetDocuments] : Asynchronous wait failed: Exceeded timeout of 1 seconds, with unfulfilled expectations: "Add or replace Movies document".

```

## Important

This draft is buggy when meilisearch does not return an error (i.e if the api link does not exist). In which case it crashes because I try to decode it as a meiliSearch Error here:
```swift
// Request.swift
let meiliSearchApiError = try! JSONDecoder().decode(MSErrorResponse.self, from: data!)
```
